In [55]:
import pandas as pd
df=pd.read_csv("data selection and filtering.csv")
df

Actor                      Film  Year            Genre  \
0       Shah Rukh Khan                   Pathaan  2023           Action   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
2           Aamir Khan                    Dangal  2016        Biography   
3        Ranbir Kapoor                Brahmastra  2022          Fantasy   
4        Ranveer Singh                 Padmaavat  2018       Historical   
5   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
6        Rajkummar Rao                     Stree  2018    Horror Comedy   
7       Hrithik Roshan                       War  2019           Action   
8         Akshay Kumar                Good Newwz  2019           Comedy   
9        Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
10        Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
11       Vicky Kaushal  Uri: The Surgical Strike  2019           Action   

    BoxOffice(INR Crore)  IMDb  
0                   1050   7.2  
1                    565   6.0  
2                   2024   8.4  
3                    431   5.6  
4                    585   7.0  
5                    111   8.3  
6                    180   7.5  
7                    475   6.5  
8                    318   7.0  
9                    266   5.9  
10                   201   6.1  
11                   342   8.2

<h1>Data Transformation</h1><br>
Once your data is clean, the next step is to <strong>reshape, reformat, and reorder </strong>it as needed for analysis.

<h3>1. Sorting & Ranking</h3>

<h5>Sort by Values</h5>

In [56]:
df2=df.sort_values("Year").copy()   #Ascending Sort
df2

Actor                      Film  Year            Genre  \
2           Aamir Khan                    Dangal  2016        Biography   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
10        Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
4        Ranveer Singh                 Padmaavat  2018       Historical   
5   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
6        Rajkummar Rao                     Stree  2018    Horror Comedy   
7       Hrithik Roshan                       War  2019           Action   
8         Akshay Kumar                Good Newwz  2019           Comedy   
11       Vicky Kaushal  Uri: The Surgical Strike  2019           Action   
3        Ranbir Kapoor                Brahmastra  2022          Fantasy   
9        Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
0       Shah Rukh Khan                   Pathaan  2023           Action   

    BoxOffice(INR Crore)  IMDb  
2                   2024   8.4  
1                    565   6.0  
10                   201   6.1  
4                    585   7.0  
5                    111   8.3  
6                    180   7.5  
7                    475   6.5  
8                    318   7.0  
11                   342   8.2  
3                    431   5.6  
9                    266   5.9  
0                   1050   7.2

In [57]:
df3=df.sort_values(["Year","IMDb"]).copy()    #Sort by multiple column ; first by year, if 2 row have same year then sort by imdb
df3

Actor                      Film  Year            Genre  \
2           Aamir Khan                    Dangal  2016        Biography   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
10        Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
4        Ranveer Singh                 Padmaavat  2018       Historical   
6        Rajkummar Rao                     Stree  2018    Horror Comedy   
5   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
7       Hrithik Roshan                       War  2019           Action   
8         Akshay Kumar                Good Newwz  2019           Comedy   
11       Vicky Kaushal  Uri: The Surgical Strike  2019           Action   
3        Ranbir Kapoor                Brahmastra  2022          Fantasy   
9        Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
0       Shah Rukh Khan                   Pathaan  2023           Action   

    BoxOffice(INR Crore)  IMDb  
2                   2024   8.4  
1                    565   6.0  
10                   201   6.1  
4                    585   7.0  
6                    180   7.5  
5                    111   8.3  
7                    475   6.5  
8                    318   7.0  
11                   342   8.2  
3                    431   5.6  
9                    266   5.9  
0                   1050   7.2

In [58]:
df.sort_values("Year", ascending=False)        #Descending Sort

Actor                      Film  Year            Genre  \
0       Shah Rukh Khan                   Pathaan  2023           Action   
3        Ranbir Kapoor                Brahmastra  2022          Fantasy   
9        Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
7       Hrithik Roshan                       War  2019           Action   
8         Akshay Kumar                Good Newwz  2019           Comedy   
11       Vicky Kaushal  Uri: The Surgical Strike  2019           Action   
4        Ranveer Singh                 Padmaavat  2018       Historical   
5   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
6        Rajkummar Rao                     Stree  2018    Horror Comedy   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
10        Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
2           Aamir Khan                    Dangal  2016        Biography   

    BoxOffice(INR Crore)  IMDb  
0                   1050   7.2  
3                    431   5.6  
9                    266   5.9  
7                    475   6.5  
8                    318   7.0  
11                   342   8.2  
4                    585   7.0  
5                    111   8.3  
6                    180   7.5  
1                    565   6.0  
10                   201   6.1  
2                   2024   8.4

<h5>Reset Index</h5>

In [59]:
df2.reset_index(drop=True, inplace=True)   #drop=True to remove new created index column; inplace=true to make changes inplace 
df2                  #  this function is used to reset the index if it is not sequential

Actor                      Film  Year            Genre  \
0           Aamir Khan                    Dangal  2016        Biography   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
2         Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
3        Ranveer Singh                 Padmaavat  2018       Historical   
4   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
5        Rajkummar Rao                     Stree  2018    Horror Comedy   
6       Hrithik Roshan                       War  2019           Action   
7         Akshay Kumar                Good Newwz  2019           Comedy   
8        Vicky Kaushal  Uri: The Surgical Strike  2019           Action   
9        Ranbir Kapoor                Brahmastra  2022          Fantasy   
10       Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
11      Shah Rukh Khan                   Pathaan  2023           Action   

    BoxOffice(INR Crore)  IMDb  
0                   2024   8.4  
1                    565   6.0  
2                    201   6.1  
3                    585   7.0  
4                    111   8.3  
5                    180   7.5  
6                    475   6.5  
7                    318   7.0  
8                    342   8.2  
9                    431   5.6  
10                   266   5.9  
11                  1050   7.2

<h5>Sort by Index</h5>

In [60]:
df3.sort_index()     #sort the DataFrame based on its index values

Actor                      Film  Year            Genre  \
0       Shah Rukh Khan                   Pathaan  2023           Action   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
2           Aamir Khan                    Dangal  2016        Biography   
3        Ranbir Kapoor                Brahmastra  2022          Fantasy   
4        Ranveer Singh                 Padmaavat  2018       Historical   
5   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
6        Rajkummar Rao                     Stree  2018    Horror Comedy   
7       Hrithik Roshan                       War  2019           Action   
8         Akshay Kumar                Good Newwz  2019           Comedy   
9        Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
10        Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
11       Vicky Kaushal  Uri: The Surgical Strike  2019           Action   

    BoxOffice(INR Crore)  IMDb  
0                   1050   7.2  
1                    565   6.0  
2                   2024   8.4  
3                    431   5.6  
4                    585   7.0  
5                    111   8.3  
6                    180   7.5  
7                    475   6.5  
8                    318   7.0  
9                    266   5.9  
10                   201   6.1  
11                   342   8.2

<h5>Ranking</h5>
The .rank() function in pandas by default gives the average rank to tied values, which can result in decimal numbers. For example, if two people share the top score, they both get a rank of 1.5. You can customize the ranking behavior using the method parameter. One useful option is method='dense', which assigns the same rank to ties but doesn’t leave gaps in the ranking sequence. This is helpful when you want a clean, consecutive ranking system without skips.

In [61]:
df2["Rank"]=df2["IMDb"].rank()    #Default: Average rank ; for eg: if two tie then rank will be 1.5 for both
df2

Actor                      Film  Year            Genre  \
0           Aamir Khan                    Dangal  2016        Biography   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
2         Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
3        Ranveer Singh                 Padmaavat  2018       Historical   
4   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
5        Rajkummar Rao                     Stree  2018    Horror Comedy   
6       Hrithik Roshan                       War  2019           Action   
7         Akshay Kumar                Good Newwz  2019           Comedy   
8        Vicky Kaushal  Uri: The Surgical Strike  2019           Action   
9        Ranbir Kapoor                Brahmastra  2022          Fantasy   
10       Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
11      Shah Rukh Khan                   Pathaan  2023           Action   

    BoxOffice(INR Crore)  IMDb  Rank  
0                   2024   8.4  12.0  
1                    565   6.0   3.0  
2                    201   6.1   4.0  
3                    585   7.0   6.5  
4                    111   8.3  11.0  
5                    180   7.5   9.0  
6                    475   6.5   5.0  
7                    318   7.0   6.5  
8                    342   8.2  10.0  
9                    431   5.6   1.0  
10                   266   5.9   2.0  
11                  1050   7.2   8.0

In [62]:
df2["Rank"]=df2["IMDb"].rank(method="dense")    #using method dense will not give rank in decimals
df2

Actor                      Film  Year            Genre  \
0           Aamir Khan                    Dangal  2016        Biography   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
2         Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
3        Ranveer Singh                 Padmaavat  2018       Historical   
4   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
5        Rajkummar Rao                     Stree  2018    Horror Comedy   
6       Hrithik Roshan                       War  2019           Action   
7         Akshay Kumar                Good Newwz  2019           Comedy   
8        Vicky Kaushal  Uri: The Surgical Strike  2019           Action   
9        Ranbir Kapoor                Brahmastra  2022          Fantasy   
10       Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
11      Shah Rukh Khan                   Pathaan  2023           Action   

    BoxOffice(INR Crore)  IMDb  Rank  
0                   2024   8.4  11.0  
1                    565   6.0   3.0  
2                    201   6.1   4.0  
3                    585   7.0   6.0  
4                    111   8.3  10.0  
5                    180   7.5   8.0  
6                    475   6.5   5.0  
7                    318   7.0   6.0  
8                    342   8.2   9.0  
9                    431   5.6   1.0  
10                   266   5.9   2.0  
11                  1050   7.2   7.0

<h3>2. Renaming Columns & Index</h3>

In [63]:
df.rename(columns={"Actor":"Hero"},inplace=True)
df

Hero                      Film  Year            Genre  \
0       Shah Rukh Khan                   Pathaan  2023           Action   
1          Salman Khan           Tiger Zinda Hai  2017           Action   
2           Aamir Khan                    Dangal  2016        Biography   
3        Ranbir Kapoor                Brahmastra  2022          Fantasy   
4        Ranveer Singh                 Padmaavat  2018       Historical   
5   Ayushmann Khurrana                 Andhadhun  2018         Thriller   
6        Rajkummar Rao                     Stree  2018    Horror Comedy   
7       Hrithik Roshan                       War  2019           Action   
8         Akshay Kumar                Good Newwz  2019           Comedy   
9        Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
10        Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
11       Vicky Kaushal  Uri: The Surgical Strike  2019           Action   

    BoxOffice(INR Crore)  IMDb  
0                   1050   7.2  
1                    565   6.0  
2                   2024   8.4  
3                    431   5.6  
4                    585   7.0  
5                    111   8.3  
6                    180   7.5  
7                    475   6.5  
8                    318   7.0  
9                    266   5.9  
10                   201   6.1  
11                   342   8.2

In [67]:
df.rename(index={0:"zero",1:"one"},inplace=True)
df

Hero                      Film  Year            Genre  \
zero      Shah Rukh Khan                   Pathaan  2023           Action   
one          Salman Khan           Tiger Zinda Hai  2017           Action   
2             Aamir Khan                    Dangal  2016        Biography   
3          Ranbir Kapoor                Brahmastra  2022          Fantasy   
4          Ranveer Singh                 Padmaavat  2018       Historical   
5     Ayushmann Khurrana                 Andhadhun  2018         Thriller   
6          Rajkummar Rao                     Stree  2018    Horror Comedy   
7         Hrithik Roshan                       War  2019           Action   
8           Akshay Kumar                Good Newwz  2019           Comedy   
9          Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
10          Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
11         Vicky Kaushal  Uri: The Surgical Strike  2019           Action   

      BoxOffice(INR Crore)  IMDb  
zero                  1050   7.2  
one                    565   6.0  
2                     2024   8.4  
3                      431   5.6  
4                      585   7.0  
5                      111   8.3  
6                      180   7.5  
7                      475   6.5  
8                      318   7.0  
9                      266   5.9  
10                     201   6.1  
11                     342   8.2

To rename all columns:

In [70]:
df.columns=["Star","Cinema","Saal","Type","Collection","Rating"]
df

Star                    Cinema  Saal             Type  \
zero      Shah Rukh Khan                   Pathaan  2023           Action   
one          Salman Khan           Tiger Zinda Hai  2017           Action   
2             Aamir Khan                    Dangal  2016        Biography   
3          Ranbir Kapoor                Brahmastra  2022          Fantasy   
4          Ranveer Singh                 Padmaavat  2018       Historical   
5     Ayushmann Khurrana                 Andhadhun  2018         Thriller   
6          Rajkummar Rao                     Stree  2018    Horror Comedy   
7         Hrithik Roshan                       War  2019           Action   
8           Akshay Kumar                Good Newwz  2019           Comedy   
9          Kartik Aaryan         Bhool Bhulaiyaa 2  2022    Horror Comedy   
10          Varun Dhawan     Badrinath Ki Dulhania  2017  Romantic Comedy   
11         Vicky Kaushal  Uri: The Surgical Strike  2019           Action   

      Collection  Rating  
zero        1050     7.2  
one          565     6.0  
2           2024     8.4  
3            431     5.6  
4            585     7.0  
5            111     8.3  
6            180     7.5  
7            475     6.5  
8            318     7.0  
9            266     5.9  
10           201     6.1  
11           342     8.2

<h3>3. Changing Column Order</h3>

Just pass a new list of column names:

In [71]:
df=df[["Cinema","Star","Saal","Type","Collection","Rating"]]
df

Cinema                Star  Saal             Type  \
zero                   Pathaan      Shah Rukh Khan  2023           Action   
one            Tiger Zinda Hai         Salman Khan  2017           Action   
2                       Dangal          Aamir Khan  2016        Biography   
3                   Brahmastra       Ranbir Kapoor  2022          Fantasy   
4                    Padmaavat       Ranveer Singh  2018       Historical   
5                    Andhadhun  Ayushmann Khurrana  2018         Thriller   
6                        Stree       Rajkummar Rao  2018    Horror Comedy   
7                          War      Hrithik Roshan  2019           Action   
8                   Good Newwz        Akshay Kumar  2019           Comedy   
9            Bhool Bhulaiyaa 2       Kartik Aaryan  2022    Horror Comedy   
10       Badrinath Ki Dulhania        Varun Dhawan  2017  Romantic Comedy   
11    Uri: The Surgical Strike       Vicky Kaushal  2019           Action   

      Collection  Rating  
zero        1050     7.2  
one          565     6.0  
2           2024     8.4  
3            431     5.6  
4            585     7.0  
5            111     8.3  
6            180     7.5  
7            475     6.5  
8            318     7.0  
9            266     5.9  
10           201     6.1  
11           342     8.2

You can also move one column to the front:

In [74]:
cols=["Type"] + [col for col in df.columns if col != "Type"]
df=df[cols]
df

Type                    Cinema                Star  Saal  \
zero           Action                   Pathaan      Shah Rukh Khan  2023   
one            Action           Tiger Zinda Hai         Salman Khan  2017   
2           Biography                    Dangal          Aamir Khan  2016   
3             Fantasy                Brahmastra       Ranbir Kapoor  2022   
4          Historical                 Padmaavat       Ranveer Singh  2018   
5            Thriller                 Andhadhun  Ayushmann Khurrana  2018   
6       Horror Comedy                     Stree       Rajkummar Rao  2018   
7              Action                       War      Hrithik Roshan  2019   
8              Comedy                Good Newwz        Akshay Kumar  2019   
9       Horror Comedy         Bhool Bhulaiyaa 2       Kartik Aaryan  2022   
10    Romantic Comedy     Badrinath Ki Dulhania        Varun Dhawan  2017   
11             Action  Uri: The Surgical Strike       Vicky Kaushal  2019   

      Collection  Rating  
zero        1050     7.2  
one          565     6.0  
2           2024     8.4  
3            431     5.6  
4            585     7.0  
5            111     8.3  
6            180     7.5  
7            475     6.5  
8            318     7.0  
9            266     5.9  
10           201     6.1  
11           342     8.2

<h1>Summary</h1>
<ul>
    <li>Sort, rank, and rename to prepare your data</li>
    <li>Reordering and reshaping are key for EDA and visualization</li>
</ul>

<h1>Melt() & Pivot()</h1>

<h3>1. melt() — Wide to Long</h3>
The melt() method in Pandas is used to unpivot a DataFrame from wide format to long format.<br>
Syntax:

Parameters:<ul>
    <li>id_vars: The columns that you want to keep fixed (these columns will remain as identifiers).</li>
    <li>value_vars: The columns you want to unpivot (the ones you want to "melt" into a single column).</li>
    <li>var_name: The name to use for the new column that will contain the names of the melted columns (default is 'variable').</li>
    <li>value_name: The name to use for the new column that will contain the values from the melted columns (default is 'value').</li>
    <li>col_level: Used for multi-level column DataFrames.</li>
</ul>

In [81]:
data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Math': [85, 78, 92],
    'Science': [90, 82, 89],
    'English': [88, 85, 94]
}
df = pd.DataFrame(data)
df   #Example df


Name  Math  Science  English
0    Alice    85       90       88
1      Bob    78       82       85
2  Charlie    92       89       94

In [78]:
df.melt(id_vars="Name",value_vars=["Math","Science","English"],var_name="Subject",value_name="Score")

Name  Subject  Score
0    Alice     Math     85
1      Bob     Math     78
2  Charlie     Math     92
3    Alice  Science     90
4      Bob  Science     82
5  Charlie  Science     89
6    Alice  English     88
7      Bob  English     85
8  Charlie  English     94

Why Use melt()?
<ul>
    <li>Data normalization: Helps in transforming data for statistical modeling and data visualization.</li>
    <li>Pivot tables: Many times, plotting functions or statistical models work better with long-format data.</li>
</ul>

<h3>2. pivot() — Long to Wide</h3>

The pivot() function in Pandas is used to reshape data, specifically to turn long-format data into wide-format data. <br>
This is the reverse operation of melt().<br>
Syntax:

Parameters:
<ul>
    <li>index: The column whose unique values will become the rows of the new DataFrame.</li>
    <li>columns: The column whose unique values will become the columns of the new DataFrame.</li>
    <li>values: The column whose values will fill the new DataFrame. These will become the actual data (values in the table).</li>
</ul>

In [84]:
data = {
    'Name': ['Alice', 'Alice', 'Alice',
             'Bob', 'Bob', 'Bob',
             'Charlie', 'Charlie', 'Charlie'],
    'Subject': ['Math', 'Science', 'English',
                'Math', 'Science', 'English',
                'Math', 'Science', 'English'],
    'Score': [85, 90, 88,
              78, 82, 85,
              92, 89, 94]
}
df2 = pd.DataFrame(data)
df2    #Example df

Name  Subject  Score
0    Alice     Math     85
1    Alice  Science     90
2    Alice  English     88
3      Bob     Math     78
4      Bob  Science     82
5      Bob  English     85
6  Charlie     Math     92
7  Charlie  Science     89
8  Charlie  English     94

In [89]:
df.pivot(index="Name", columns="Subject", values="Score")

Subject  English  Math  Science
Name                           
Alice         88    85       90
Bob           85    78       82
Charlie       94    92       89

Why use pivot()?
<ul>
    <li>Better data structure: It makes data easier to analyze when you have categories that you want to split into multiple columns.</li>
    <li>Easier visualization: Often, you want to represent data in a format where categories are split across columns (for example, when creating pivot tables for reporting).</li>
    <li>Aggregating data: You can perform aggregations (like sum, mean, etc.) to group values before pivoting.</li>
</ul>

<h3>Important Notes:</h3>
<strong>1. Duplicate Entries:</strong> If you have multiple rows with the same combination of index and columns, pivot() will raise an error. <br>
In such cases, you should use pivot_table() (which can handle duplicate entries by aggregating them).

In [91]:
data = {
    'Name': ['Alice', 'Alice', 'Alice', 'Bob', 'Bob'],
    'Subject': ['Math', 'Math', 'Science', 'Math', 'Math'],
    'Score': [85, 80, 90, 78, 82]
}
df = pd.DataFrame(data)
df

Name  Subject  Score
0  Alice     Math     85
1  Alice     Math     80
2  Alice  Science     90
3    Bob     Math     78
4    Bob     Math     82

We can use pivot_table() to aggregate values (e.g., taking the mean for duplicate entries):

In [92]:
df.pivot_table(index="Name", columns="Subject", values="Score", aggfunc="mean")

Subject  Math  Science
Name                  
Alice    82.5     90.0
Bob      80.0      NaN

In this case, the Math score for Alice is averaged (85 + 80) / 2 = 82.5. If a cell is empty, it means there was no value for that combination.


<h1>Summary:</h1>
<ul>
    <li>Use melt() to go long, pivot() to go wide</li>
    <li>pivot() is used to turn long-format data into wide-format by spreading unique column values into separate columns.</li>
    <li>If there are duplicate values for a given combination of index and columns, you should use pivot_table() with an aggregation function to handle the duplicates.</li>
</ul>